packages/internal/src/utils.js

```diff
@@ -411,37 +411,50 @@ export const assertAllDefined = obj => {
   }
 };
 
-const neverDone = harden({ done: false, value: null });
+/** @type {IteratorResult<undefined, never>} */
+const neverDone = harden({ done: false, value: undefined });
+
+/** @type {IteratorResult<never, void>} */
+const alwaysDone = harden({ done: true, value: undefined });
 
-/** @type {AsyncIterable<null>} */
 export const forever = asyncGenerate(() => neverDone);
 
 /**
- * @param {() => unknown} boolFunc
- * `boolFunc`'s return value is used for its truthiness vs falsiness.
+ * @template T
+ * @param {() => T} produce
+ * `produce`'s return value is used for its truthiness vs falsiness.
  * IOW, it is coerced to a boolean so the caller need not bother doing this
  * themselves.
- * @returns {AsyncIterable<null>}
+ * @returns {AsyncIterable<Awaited<T>>}
  */
-export const whileTrue = boolFunc =>
-  asyncGenerate(() =>
-    harden({
-      done: !boolFunc(),
-      value: null,
-    }),
-  );
+export const whileTrue = produce =>
+  asyncGenerate(async () => {
+    const value = await produce();
+    if (!value) {
+      return alwaysDone;
+    }
+    return harden({
+      done: false,
+      value,
+    });
+  });
 
 /**
- * @param {() => unknown} boolFunc
- * `boolFunc`'s return value is used for its truthiness vs falsiness.
+ * @template T
+ * @param {() => T} produce
+ * `produce`'s return value is used for its truthiness vs falsiness.
  * IOW, it is coerced to a boolean so the caller need not bother doing this
  * themselves.
- * @returns {AsyncIterable<null>}
+ * @returns {AsyncIterable<Awaited<T>>}
  */
-export const untilTrue = boolFunc =>
-  asyncGenerate(() =>
-    harden({
-      done: !!boolFunc(),
-      value: null,
-    }),
-  );
+export const untilTrue = produce =>
+  asyncGenerate(async () => {
+    const value = await produce();
+    if (value) {
+      return alwaysDone;
+    }
+    return harden({
+      done: false,
+      value,
+    });
+  });
```

Comment:
For `whileTrue`, the enhancement to provide the truthy awaited value as the non-final iteration value seems plausible enough. After all, falsy values are less often interesting in themselves than are truthy values. But here, you're doing the opposite --- providing only the falsy values, because only those are non-final. Given that, I think it would be better to either 
   * revert to the previous behavior, where only the truthiness of the (awaited) value is significant. Here, that would simply be

```suggestion
export const untilTrue = produce =>
  asyncGenerate(async () => {
    const value = await produce();
    if (value) {
      return alwaysDone;
    }
    return neverDone;
  });
```

Btw, the fact that `neverDone` could plausibly be used for a single non-final iteration that does eventually terminate indicates that `neverDone` might not be the right name. How about `notDone`?

--- 
or, 
   * provide the actual (awaited) value in both cases

```suggestion
export const untilTrue = produce =>
  asyncGenerate(async () => {
    const value = await produce();
    if (value) {
      return harden({
        done: true,
        value
      });
    }
    return harden({
      done: false,
      value,
    });
  });
```

I do understand that `for await of` loops can never see the final values. But our imagined uses would rarely look at non-final values either, and not all our uses are necessarily via `for await of` loops.

Reply:
I changed `neverDone` to `notDone` in the rest of this file.
> provide the actual (awaited) value in both cases

I like this, so I did it.